### PR TITLE
Skeleton and Skeleton inspector low-level changes

### PR DESCRIPTION
**Pull request by TwistedTwigleg**
Some low-level changes to both the Skeleton3D and the awesome new Skeleton3D inspector!

Changes:
- [x] ~~Made the `Bone` struct in `Skeleton3D` public so it can be used if/when needed for the gizmo/sub-gizmo system.~~ Removed! It can be added again later.
- [x] Added helper functions for converting transforms from bone space to world space, and vise versa.
- [x] Updated the `Skeleton3D` class reference with the new functions, and filled out some of the existing functions.
- [x] ~~(Tried exposing the `Bone` struct to GDScript and found it was clunky to use and the functions provided in Skeleton were better suited. A few commits are related to this, but the code added was removed afterwards)~~
- [x] Changed the icon used in the `Skeleton3D` inspector to the BoneAttachement, so the bones shown in the inspector use the bone icon.
- [x] Changed the `Skeleton3D` inspector to use `EditorPropertyTransform` instead of EditorSpinSliders for the transform/matrix.
- [x] Created a new function in `EditorPropertyTransform` that allows the UI to be set with a passed-in transform.
- [x] Use `EditorPropertyVector3` for the translation section of the `Skeleton3D` inspector
- [x] Use `EditorPropertyVector3` for the rotation section of the `Skeleton3D` inspector
- [x] Use `EditorPropertyVector3` for the scale section of the `Skeleton3D` inspector
- [x] Placed the Transform/Matrix into a sub-section, so it mirrors the style used in `Node3D`.
  * I thought about putting the position, rotation, and scale also in sub-sections, but I'm not sure there would be much benefit. Usability wise, I think just having the Transform/Matrix in a sub-section is probably fine.
- [x] Fix Travis error(s)

Additional PR ideas:
* Add per-bone inspector visibility.
  * This should make it easier to manipulate bones in meshes that have 24+ bones in them, as the unnecessary/infrequently used bones can be hidden away until needed. Might be a tad complex to implement.
  * Note: this is just for the inspector. The bones would still be visible on the 3D model. It is just for ease of use when working on 3D models with a lot of bones.
* Make a list of child bones in the Bone struct so it is easier to process parent-child relations.
* Make a gizmo/sub-gizmo system for manipulating bones
  * Not sure on the complexity of this. Hard to know what sub-tasks will be needed until I start working on it.
* Add a search/filter box so it is easier to find bones.
  * Thanks to @slapin for the suggestion!

~**Note**: the commits are a bit messy, but it is because I didn't want to accidentally lose work/changes as I go along. I will clean them up if needed, or they can be squashed into a single commit upon being merged. Either way is fine with me.~ Fixed!

*Disclaimer: This work is part of the GSOC 2020 program.*

### CODE REVIEW

**Comment by fire** [comment]
Was the Bone struct used or not used? Merging the commits helps it be clear.

**Comment by TwistedTwigleg** [comment]
The BoneStruct3D ultimately ended up not being used.
The Bone struct is also, currently, unused outside of the Skeleton3D class, though I *think* it will be needed later for the gizmo/sub-gizmo.

I probably should squish, merge, or drop commits, but I didn't want to accidentally lose work, which unfortunately I have done before. I can probably merge some of the commits together though, at least so the Skeleton and Skeleton inspector stuff is separate.

File: doc/classes/Skeleton3D.xml
Comment (fire) [inline doc/classes/Skeleton3D.xml:40]: Which world transform is it?

* The "global transform" from the root node to the bone
* The "world transform" from the skeleton node to the bone. It's probably this one.

Comment (TwistedTwigleg) [reply doc/classes/Skeleton3D.xml:40]: The latter. I will revise this so it is more clearly stated :+1:

File: editor/plugins/skeleton_3d_editor_plugin.cpp
Comment (fire) [inline editor/plugins/skeleton_3d_editor_plugin.cpp:71]: At some point it would be good to convert their to the new callable interface, but I didn't because it made backporting to 3.2 harder.

Will you backport this to 3.2?

Comment (TwistedTwigleg) [reply editor/plugins/skeleton_3d_editor_plugin.cpp:71]: Yeah, using the callable interface might have been better, but I figured it was best to follow the style already used in the class (and it was what I am already familiar with, so a win-win!)

I can probably backport the changes to Godot 3.2 once this PR is merged. At very least, I can backport the EditorPropertyTransform and EditorPropertyVector3 changes.

**Comment by fire** [comment]
> Add per-bone inspector visibility. 

Take care that you handle the children of the bone case. Could be a small pr.

> Make a list of child bones in the Bone struct so it is easier to process parent-child relations.

Take care you keep one place as the singular definition of parent / children.

> Make a gizmo/sub-gizmo system for manipulating bones 

Could be a small separate pr.

File: editor/plugins/skeleton_3d_editor_plugin.cpp
Comment (fire) [reply editor/plugins/skeleton_3d_editor_plugin.cpp:71]: I'll try to get https://github.com/godotengine/godot/pull/39090 merged, or alternatively you can take over the pr.

File: doc/classes/Skeleton3D.xml
Comment (fire) [inline doc/classes/Skeleton3D.xml:212]: This one is a global transform right?

A spatial is relation to the root node. While the bone is in relation to the skeleton or just the bone.

Comment (fire) [reply doc/classes/Skeleton3D.xml:212]: Or did you write spatial as a shorthand for Transform3D.

Comment (fire) [reply doc/classes/Skeleton3D.xml:212]: The custom pose is a replacement for the bone rest.

**Comment by TwistedTwigleg** [comment]
> > Add per-bone inspector visibility.
>
>  Take care that you handle the children of the bone case. Could be a small pr.
>
> > Make a list of child bones in the Bone struct so it is easier to process parent-child relations.
>
> Take care you keep one place as the singular definition of parent / children.

Yup! In both cases, I think it would be better as a small, separate PR. This PR should probably just be focused on low-level, minor adjustments and changes. I'll edit the OP accordingly.

> > Make a gizmo/sub-gizmo system for manipulating bones
>
> Could be a small separate pr.

Indeed! That would also make it easier to backport to Godot 3.2, assuming it is possible and does not require substantial changes. I'll also adjust this in the OP.

Thanks you for the feedback! I'll make the changes soon!

File: editor/plugins/skeleton_3d_editor_plugin.cpp
Comment (TwistedTwigleg) [reply editor/plugins/skeleton_3d_editor_plugin.cpp:71]: Whatever works best for you works for me.

**Comment by fire** [comment]
Some cicd failues here https://travis-ci.org/github/godotengine/godot/jobs/692253680

**Comment by TwistedTwigleg** [comment]
The clang format issues should be fixed now! It took me a bit to get clang-format working properly, and then applied correctly, but everything seems to be good now.

**Comment by fire** [comment]
You can save a backup branch, but this one should be squashed because it has code in one commit that is deleted in the other.

**Comment by fire** [comment]
Note: this is code review, I haven't compiled the engine with these changes yet. If you can execute something similar to this test plan. My plan is to take a look during the weekend.

    Able to select bones from the tree
    Able to rearrange the bone tree order and undo works.
    Playing animations will change the pose info
    Can key animations
    Can edit the bone in the three types: pose, rest and custom
    Can move the input boxes by 0.001.
    None of the input boxes are static / stuck.

**Comment by fire** [comment]
If you aren't using the bone struct currently, I don't think it should be exposed.

image

> #2: The problem has to exist

https://docs.godotengine.org/en/stable/community/contributing/best_practices_for_engine_contributors.html

File: scene/3d/skeleton_3d.h
Comment (fire) [inline scene/3d/skeleton_3d.h:193]: Are these being used in this pr directly?

Comment (fire) [reply scene/3d/skeleton_3d.h:193]: You're free to add it your subgizmos pr though.

Comment (fire) [reply scene/3d/skeleton_3d.h:193]: Exposing would require the struct to be a resource with all the wrappers and gdscript bindings.

Comment (TwistedTwigleg) [reply scene/3d/skeleton_3d.h:193]: > Exposing would require the struct to be a resource with all the wrappers and gdscript bindings.

I know, I went down that route. Adding the custom resource with all the wrappers and GDScript bindings was 3-4 of the squished commits! I ultimately found it was not worth the extra complexity and using the Struct was more clunky than just using the functions in Skeleton.

I can probably remove those functions and move the Bone struct back to a private struct now. As you said, if the sub-gizmo needs it, it can be changed again. I'll make a PR with these changes soon.

Comment (fire) [reply scene/3d/skeleton_3d.h]: The last pr should be combined with the pr that made the change so it has a change commit. Note that history is lost here.

Comment (TwistedTwigleg) [reply scene/3d/skeleton_3d.h]: I tried, but the rebase gets all messed up when I try to move the commit to override the first commit. Is there a way to rebase and reorder the commits without the log getting messed up? I tried moving the actual `fixup`/`squash` line so it was under the initial commit, and it didn't work. There was merge conflicts for every single commit, even for those that would otherwise be unaffected.

Comment (fire) [reply scene/3d/skeleton_3d.h]: It's tricky, and that's why I squish all of my commits together, because it makes porting to 3.2 easier. (One cherry pick and then fix conflicts.)

**Comment by fire** [comment]
I would recommend squashing so there isn't a extra commit that would confuse code readers.

After that, I can add my approval and wait for @akien-mga to merge.

I don't recommend adding more enhancements to this pr.

File: doc/classes/Skeleton3D.xml
Comment (fire) [inline doc/classes/Skeleton3D.xml:212]: Typo:

lkie -> like.

File: doc/classes/Skeleton3D.xml
Comment (fire) [inline doc/classes/Skeleton3D.xml:269]: Second typo.

**Comment by TwistedTwigleg** [comment]
> I would recommend squashing so there isn't a extra commit that would confuse code readers.

Will do! I am thinking it would be best to squish everything into a single commit (and rework the commit message so it makes sense) for easier porting to Godot 3.2

> I don't recommend adding more enhancements to this pr.

Probably best, though I have been working on exposing the global pose override as well, since it could be helpful for debugging and would compete the set of bone transforms shown in the inspector.
**Should I leave adding the bone global override pose for another PR?** I have not made a commit yet, so it should be simple enough to split the work off into another PR.

**Comment by TwistedTwigleg** [comment]
@fire - I squashed everything into one commit and fixed the typos! Everything should be good now. 

Please let me know if anything is out of place, and thanks so much for the help, I really appreciate it!

**Review by fire — APPROVED** [review]

**Comment by slapin** [comment]
Anybody working with inspector changes - please add bone search option. It is impossibly hard to work with realistic bone counts (120+).

**Comment by fire** [comment]
@slapin Can you write a Godot Engine proposal on how bone search option should look? Make some mockups? That would help me.

**Comment by slapin** [comment]
Just a box. Nothing fancy. Just to limit number of displayed bones. This
stuff is already used in Godot, so no reason for any new design
(see props filter in inspector).
And no proposal due to too low value/time and too much communication noise.

**Comment by TwistedTwigleg** [comment]
It should be possible to make a filter, that will only show bones with the inputted name. On the loop that adds the bones to the Tree node, a condition could be added that would discard results that do not have the inputted filter/search string in it. It wouldn't be a search exactly, but I think it would achieve the same effect (or most of it) without really needing too many changes to the code for the inspector. Though, special care may need to be taken for parent/child relations and how they are displayed.

Probably beyond the scope of this PR though. I'll add it to the list of additional PR ideas list in the opening post.

File: scene/3d/skeleton_3d.cpp
Comment (fire) [inline scene/3d/skeleton_3d.cpp:862]: How is this different from get_global_transform() * p_bone_transform?

Also this appears to be the global transform of the skeleton and the local bone transform? 

This isn't the transform needed to go from the skeleton node to the bone from the root spatial node transform.

Comment (fire) [reply scene/3d/skeleton_3d.cpp:862]: Also, above information can be calculated when the process order of the bone struct is calculated. then the global transform of the skeleton can be fetched.

TL;DR, why isn't this in the bone struct?

Comment (TwistedTwigleg) [reply scene/3d/skeleton_3d.cpp:862]: > How is this different from get_global_transform() * p_bone_transform?

Hmm, it might not be any different, thinking about it. If that snippet works, then that would be ideal. I will need to test! I am all for using simpler, easier to understand code :+1: 

> Also this appears to be the global transform of the skeleton and the local bone transform?

Here's what the function is intended to be used for, which hopefully can help better explain more about it:

The function is for taking a bone global pose (`global_pose`, `global_pose_override`, or one of the local poses converted to a global pose) and returns a transform that is the bone's position in the scene. `bone_transform_to_world_transform` is so you can do computations with the transform as if it is a global transform received from a Node3D. Then, when you are finished using the transform in the scene and want to apply it to the bone transform, you can use `world_transform_to_bone_transform` to get the transform as a global pose, when then you can apply.

Hopefully that helps clear things up a bit. It is kinda hard to explain, but it can be extremely useful to have when you need it.
I"m not totally happy with the naming though. I'd prefer to change the name(s) of the functions in #39353 rather than here though, just because said PR builds on top of this one. I am thinking a more descriptive name would be something like `global_pose_to_world_transform` and `world_transform_to_global_pose`.

> TL;DR, why isn't this in the bone struct?

A few reasons:

The first reason is that the helper functions may be called using multiple different transforms, like the newly introduced `local_pose_override` in my draft PR or the existing `global_pose_override`. Ultimately, it would probably only ever by the two override poses that ever get used with these functions, but then the Bone struct would need an additional two transforms.

The second reason is that including it in the Bone struct would increase the computations for those using just the Skeleton3D node. If possible, I figure it is best to try and make the IK code mostly separate and limit the performance impact it could have for those not using it.

The third reason is that, with how the IK needs to be run, it would be recalculated more often than needed as the bones in the chain are iteratively updated. Keeping it a separate function means it is only used when needed. Especially for iterative IK, like FABRIK, CCDIK, etc, the cost could add up quickly when calculating it every time one of the bones changes. Keeping it out of the Bone struct avoids this issue entirely.

Comment (fire) [reply scene/3d/skeleton_3d.cpp:862]: I'm not sure which is faster, keeping the data in the bones struct vs going from the bone all the way to the parent bone every time the function is called. 

Probably need to test it. 

We thought that recreating all the transforms from the bone all the way to the bone chain root would be expensive, but it's really better done as a bench mark with 300-500 bones.

You do need to iterate to the root of the bone chain operation, but as mentioned it's undecided keeping the data as part of the struct is slower than recomputing every function call.

**Comment by TwistedTwigleg** [comment]
Simplified the `bone_transform_to_world_transform` and `world_transform_to_bone_transform` functions based on feedback from @fire (Thanks!)
Updated this PR (locally) to be using a newer version of the master branch.

**Comment by slapin** [comment]
caching transform is actually faster for larger skeletons

**Comment by TwistedTwigleg** [comment]
With IK though, and how the functions are going to be used, they will only be used on tops a couple dozen bones. I *highly* doubt anyone would use IK and have it operate on hundreds of bones, I cannot really see any uses. The only thing I can think of that might have several dozen bones is wings, like big dragon wings, and tails, but that would be in the dozens at most.

However, IK and the functions will likely be used on skeletons that lots, even hundreds, of bones, it just won't be operating on them directly. In this case, adding it to the Bone struct would cause *all* of these bones to run the calculation, not just the ones that are being operated on by IK. In this case, I think the performance loss would be greater than the potential gain. I think there would be more unused calculations being performed than used for large skeletons. For smaller skeletons, I would expect the performance to be more or less the same either way.

(Also: this is not counting having to calculate this every time the bone transform changes, needed for iterative IK, which also has a performance cost)

_______

Regardless, we would really need to run benchmarks to know for sure. I think it is best not to prematurely optimize and leave it as a function for now. It should be easy enough to change if, after testing, it is discovered that caching in the Bone struct is more performance friendly.

**Comment by slapin** [comment]
caching will let you pass through "clean" bones and touching only
"dirty" ones and I don't see much strain
on CPU for update if done right.

**Comment by fire** [comment]
Large numbers of bones being directly used/solved by IK is part of my use case design, but it should be fast for the default case too.

> this is not counting having to calculate this every time the bone transform changes

Why not count it?

How do you want to design the benchmark? Both of our reasoning arguments seem flimsy.

**Comment by slapin** [comment]
Also current IK can't do sparse chains (like ones in Unity's Mecanim)
so cache will still be relevant
in reducing numer of calculations.

**Comment by slapin** [comment]
Yeah full-body IK is quite common these days (requires sparse chain support
though) and caching helps saving CPU time

**Review by JFonS — APPROVED** [review]
Looks good!

File: doc/classes/Skeleton3D.xml
Comment (akien-mga) [inline doc/classes/Skeleton3D.xml:151]:
```suggestion
				Returns whether the bone rest for the bone at [code]bone_idx[/code] is disabled.
```

File: doc/classes/Skeleton3D.xml
Comment (akien-mga) [inline doc/classes/Skeleton3D.xml:216]: There should not be empty lines, nor cosmetic line breaks (line on line 218) in the XML. Each line == a paragraph.

Comment (akien-mga) [reply doc/classes/Skeleton3D.xml:216]: This applies to several of the descriptions below.

File: editor/plugins/skeleton_3d_editor_plugin.cpp
Comment (akien-mga) [reply editor/plugins/skeleton_3d_editor_plugin.cpp:71]: Yes please use `connect_mp` here instead of keeping the compat method. That removes the need to bind the method just to connect to it.

File: editor/plugins/skeleton_3d_editor_plugin.cpp
Comment (akien-mga) [inline editor/plugins/skeleton_3d_editor_plugin.cpp:193]: Can `Basis()` stay on a single line? With `clang-format` this isn't really more readable than a one liner.

File: scene/3d/skeleton_3d.cpp
Comment (akien-mga) [inline scene/3d/skeleton_3d.cpp:862]: Ensure that there is a separation line between all function definitions.

File: editor/plugins/skeleton_3d_editor_plugin.cpp
Comment (akien-mga) [inline editor/plugins/skeleton_3d_editor_plugin.cpp:545]: I guess this should be removed?

File: editor/plugins/skeleton_3d_editor_plugin.cpp
Comment (akien-mga) [inline editor/plugins/skeleton_3d_editor_plugin.cpp:546]: Shouldn't this be `Ref<Texture2D>`?

**Comment by TwistedTwigleg** [comment]
@JFonS - Thanks for the review!
@akien-mga - I think I fixed everything brought up in your review. It should be good to go now :)

File: doc/classes/Skeleton3D.xml
Comment (akien-mga) [inline doc/classes/Skeleton3D.xml:41]: Empty line here.

Comment (TwistedTwigleg) [reply doc/classes/Skeleton3D.xml]: Sorry about that! I thought I got them all. It should be fixed now.

File: editor/plugins/skeleton_3d_editor_plugin.cpp
Comment (TwistedTwigleg) [reply editor/plugins/skeleton_3d_editor_plugin.cpp:71]: @akien-mga - Sorry, didn't see this comment until now! I tried changing the `connect_compact` to `connect_mp`, but there does not appear to be a `connect_mp` function, at least not in the version of master I used last.

Was the function just recently added, like in the last week or two? If not, how do I use the function? I did a project wide search, but `connect_mp` didn't show any results.

Comment (akien-mga) [reply editor/plugins/skeleton_3d_editor_plugin.cpp:71]: Sorry that was a typo. It's `connect`, and the callable argument should use `callable_mp(this, &ClassName::method_name)`.

**Comment by TwistedTwigleg** [comment]
I converted all of the `connect_compact` function calls to just `connect`, using the format @akien-mga provided. I went through and testing all of the signal calls changed, and everything seems to be working. :+1:

File: editor/plugins/skeleton_3d_editor_plugin.cpp
Comment (akien-mga) [inline editor/plugins/skeleton_3d_editor_plugin.cpp:225]: Since you use `callable_mp` which refers to the function pointer directly, you don't need to bind the method. Similarly, all three other pre-existing bindings can be removed (and thus the whole `_bind_methods` override).

Comment (TwistedTwigleg) [reply editor/plugins/skeleton_3d_editor_plugin.cpp]: I removed the `_bind_methods` override since it's no longer needed. :+1:

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
